diff --git a/field_permissions.module b/field_permissions.module
index 84c5e78760c5340f09e60b5d3b21b1e6df071f27..48cf0a78ddb05ffade7544733042ffd5190470ac 100644
--- a/field_permissions.module
+++ b/field_permissions.module
@@ -47,7 +47,7 @@ function field_permissions_entity_field_access($operation, FieldDefinitionInterf
  * Implements hook_jsonapi_entity_field_filter_access().
  */
 function field_permissions_jsonapi_entity_field_filter_access(FieldDefinitionInterface $field_definition, AccountInterface $account) {
-  if (!$field_definition->isDisplayConfigurable('view')) {
+  if (!$field_definition->isDisplayConfigurable('view') || !is_a($field_definition->getFieldStorageDefinition(), '\Drupal\field\FieldStorageConfigInterface')) {
     return AccessResult::neutral();
   }
 
