diff --git a/src/OfficeHoursFormatterTrait.php b/src/OfficeHoursFormatterTrait.php
index 5f1aeb7..b1188cb 100644
--- a/src/OfficeHoursFormatterTrait.php
+++ b/src/OfficeHoursFormatterTrait.php
@@ -550,6 +550,10 @@ trait OfficeHoursFormatterTrait {
           break;
       }
 
+      // Ensure empty comments are filtered out to prevent adjacent commas
+      // from being displayed to the end-user.
+      $info['comments'] = array_filter(array_map('trim', $info['comments']));
+
       // Concatenate the comment lines.
       $info['comments'] = implode($slot_separator, $info['comments']);
       // For compressed items, remove trailing separator.
