diff --git a/src/OfficeHoursFormatterTrait.php b/src/OfficeHoursFormatterTrait.php
index 4be2a9d..9d8cb09 100644
--- a/src/OfficeHoursFormatterTrait.php
+++ b/src/OfficeHoursFormatterTrait.php
@@ -392,6 +392,7 @@ trait OfficeHoursFormatterTrait {
       if ($field_settings['comment'] == 2) {
         $info['comments'] = array_map('t', $info['comments']);
       }
+      $info['comments'] = array_filter(array_map('trim', $info['comments']));
       $info['comments'] = ($field_settings['comment'])
           ? implode($slot_separator, $info['comments'])
           : '';
